src/encdatavault_fmt_plug.c

```diff
@@ -0,0 +1,434 @@
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // Encrypted header
+		goto err;
+	if (hexlenl(p, &extra) != ENC_SIG_SIZE * 2 || extra)
+		goto err;
+	MEM_FREE(keeptr);
+	return 1;
+
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs;
+	int i;
+	char *p = ciphertext, *ctcopy, *keeptr;
+
+	memset(&cs, 0, sizeof(cs));
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	p = strtokm(ctcopy, "$");
+	cs.version = atoi(p);
+	p = strtokm(NULL, "$");
+	cs.algo_id = atoi(p);
+	p = strtokm(NULL, "$");
+	for (i = 0; i < ENC_NONCE_SIZE; i++)
+		cs.iv[i] =
+		    (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i +
+		            1])];
+
+	// Init AES CTR counter
+	cs.iv[ENC_IV_SIZE - 1] = 1;
+
+	p = strtokm(NULL, "$");
+	for (i = 0; i < ENC_SIG_SIZE; i++)
+		cs.encrypted_data[i + 4] =
+		    (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i +
+		            1])];
+
+	// Keychain for version 3
+	if (cs.version == 3) {
+		p = strtokm(NULL, "$");
+		for (i = 0; i < ENC_KEYCHAIN_SIZE; i++)
+			cs.keychain[i] =
+			    (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 *
+			            i + 1])];
+	}
+
+	MEM_FREE(keeptr);
+	return (void *)&cs;
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index;
+	int nb_keys;
+
+	if (any_cracked) {
+		memset(cracked, 0, cracked_size);
+		any_cracked = 0;
+	}
+
+	if (1 == cur_salt->version) {
+		nb_keys = 1 << (cur_salt->algo_id - 1);
+	} else {
+		nb_keys = 8;
+	}
+
+#ifdef _OPENMP
+	#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index++) {
+		int j;
+		MD5_CTX ctx;
+		buffer_128 kdf_out[ENC_MAX_KEY_NUM];
+		buffer_128 hash;
+		buffer_128 tmp;
+		buffer_128 ivs[ENC_MAX_KEY_NUM] = { 0 };
+		unsigned char result[ENC_KEY_SIZE * ENC_MAX_KEY_NUM] = { 0 };
+		unsigned char *result_ptr;
+
+		// Key derivation
+		tmp.u64[0] = 0;
+		tmp.u64[1] = 0;
+		MD5_Init(&ctx);
+		MD5_Update(&ctx, saved_key[index], strlen(saved_key[index]));
+		MD5_Final(hash.u8, &ctx);
+
+		for (j = 0; j < nb_keys; j++) {
+			memcpy(kdf_out[j].u8, default_salts[j], ENC_SALT_SIZE);
+		}
+
+		for (j = 1; j < ENC_DEFAULT_NUM_HASH_ITERATIONS; j++) {
+			MD5_Init(&ctx);
+			MD5_Update(&ctx, hash.u8, 16);
+			MD5_Final(hash.u8, &ctx);
+			enc_xor_block(tmp.u64, hash.u64);
+		}
+
+		for (j = 0; j < nb_keys; j++) {
+			enc_xor_block(kdf_out[j].u64, tmp.u64);
+		}
+
+		/* AES iterated CTR */
+		if (cur_salt->version == 1) {
+			memcpy(ivs[0].u8, cur_salt->iv, ENC_NONCE_SIZE);
+			for (j = 1; j < nb_keys; j++) {
+				memcpy(ivs[j].u8, cur_salt->iv, ENC_NONCE_SIZE);
+				ivs[j].u64[0] ^= kdf_out[j].u64[0];
+			}
+			enc_aes_ctr_iterated(cur_salt->encrypted_data, result,
+			                     kdf_out[0].u8, ivs, AES_BLOCK_SIZE, nb_keys, 1);
+			result_ptr = result + 4;
```

Comment:
I see no need for the separate `result_ptr` variable here and in the other place (and thus at all). You can just pass `result + 4` into the `memcmp`.

Reply:
Perhaps more importantly - `result` is kind of a two-dimensional array, but here (and in the other place) you always test the first "element", right? Is that right, and why?

Reply:
`result` contains first the decrypted keys from the keychain (128 bytes) at line 322 and then I reuse the buffer to decrypt the data which is smaller.

Reply:
> result contains first the decrypted keys from the keychain (128 bytes) at line 322 and then I reuse the buffer to decrypt the data which is smaller.

You could want to add a source code comment on that.